## Load existing dataset, collect Data from Spotify API, join the data.

In [1]:
import pandas as pd

### Load the dataset used by previous DS team

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/Lambda-School-Labs/djhelper-ds/visuals_canvas/Data/SpotifyFeatures.csv')

### Look at the data

In [3]:
df.head()

genre        artist_name                        track_name  \
0  Movie     Henri Salvador       C'est beau de faire un Show   
1  Movie  Martin & les fées  Perdu d'avance (par Gad Elmaleh)   
2  Movie    Joseph Williams    Don't Let Me Be Lonely Tonight   
3  Movie     Henri Salvador    Dis-moi Monsieur Gordon Cooper   
4  Movie       Fabien Nataf                         Ouverture   

                 track_id  popularity  acousticness  danceability  \
0  0BRjO6ga9RKCKjfDqeFgWV           0         0.611         0.389   
1  0BjC1NfoEOOusryehmNudP           1         0.246         0.590   
2  0CoSDzoNIKCRs124s9uTVy           3         0.952         0.663   
3  0Gc6TVm52BwZD07Ki6tIvf           0         0.703         0.240   
4  0IuslXpMROHdEPvSl1fTQK           4         0.950         0.331   

   duration_ms  energy  instrumentalness key  liveness  loudness   mode  \
0        99373   0.910             0.000  C#    0.3460    -1.828  Major   
1       137373   0.737             0.000  F#    0.1510    -5.559  Minor   
2       170267   0.131             0.000   C    0.1030   -13.879  Minor   
3       152427   0.326             0.000  C#    0.0985   -12.178  Major   
4        82625   0.225             0.123   F    0.2020   -21.150  Major   

   speechiness    tempo time_signature  valence  
0       0.0525  166.969            4/4    0.814  
1       0.0868  174.003            4/4    0.816  
2       0.0362   99.488            5/4    0.368  
3       0.0395  171.758            4/4    0.227  
4       0.0456  140.576            4/4    0.390

In [4]:
df.shape

(232725, 18)

In [5]:
df['genre'].value_counts()

Comedy              9681
Soundtrack          9646
Indie               9543
Jazz                9441
Pop                 9386
Electronic          9377
Children’s Music    9353
Folk                9299
Hip-Hop             9295
Rock                9272
Alternative         9263
Classical           9256
Rap                 9232
World               9096
Soul                9089
Blues               9023
R&B                 8992
Anime               8936
Reggaeton           8927
Ska                 8874
Reggae              8771
Dance               8701
Country             8664
Opera               8280
Movie               7806
Children's Music    5403
A Capella            119
Name: genre, dtype: int64

In [6]:
df['artist_name'].value_counts()[:30]

Giuseppe Verdi              1394
Giacomo Puccini             1137
Kimbo Children's Music       971
Nobuo Uematsu                825
Richard Wagner               804
Wolfgang Amadeus Mozart      800
Randy Newman                 757
Georges Bizet                701
Juice Music                  684
Johann Sebastian Bach        632
Ludwig van Beethoven         596
Hans Zimmer                  559
Gioachino Rossini            491
Chorus                       480
Howard Shore                 479
Drake                        477
Henri Salvador               474
John Williams                450
Frédéric Chopin              436
Alan Menken                  420
Chris Brown                  392
Bob Marley & The Wailers     380
Dorothée                     378
Gaetano Donizetti            377
Children Songs Company       371
Max Richter                  365
Capcom Sound Team            326
Future                       317
Yuki Hayashi                 309
Sublime                      306
Name: arti

### As we can see from the data, not a lot of popular songs are there

In [7]:
df['popularity'].value_counts()[:30]

0     6312
50    5415
53    5414
51    5401
52    5342
49    5266
48    5068
54    5013
47    4944
55    4696
46    4657
56    4632
44    4605
43    4573
45    4525
42    4472
57    4461
41    4414
40    4261
38    4253
58    4208
36    4175
39    4151
33    4096
35    4089
37    4079
32    4041
59    3981
29    3929
31    3924
Name: popularity, dtype: int64

### Let's add some 2019 - 2020 songs from Spotify by setting up Spotipy for querying the API endpoint.

In [8]:
# Load Spotify API
import spotipy
from spotipy.oauth2 import SpotifyClientCredentials
import spotipy.util as util

### Safety first. Keep your keys private

In [9]:
# First create .env file with the key values in your folder
from dotenv import load_dotenv
import os

In [10]:
load_dotenv() # load environment variables

True

In [12]:
# you can check your variables before running API client credentials.
# print("CLIENT ID:", os.environ.get("SPOTIPY_CLIENT_ID"))
# print("CLIENT SECRET:", os.environ.get("SPOTIPY_CLIENT_SECRET"))

In [13]:
client_credentials_manager = SpotifyClientCredentials()
sp = spotipy.Spotify(client_credentials_manager=client_credentials_manager)

### We will be quiring the data with the same features as the existing dataset

In [18]:
import timeit
start = timeit.default_timer()

# create empty lists where the results are going to be stored
artist_name = []
track_name = []
popularity = []
track_id = []

# the maximum query for Spotify API is 2000 with the max 
for i in range(0,2000,50):
    track_results = sp.search(q='year:2019', type='track', limit=50,offset=i)
    for i, t in enumerate(track_results['tracks']['items']):
        artist_name.append(t['artists'][0]['name'])
        track_name.append(t['name'])
        track_id.append(t['id'])
        popularity.append(t['popularity'])
      

stop = timeit.default_timer()
print ('Time to run this code (in seconds):', stop - start)

Time to run this code (in seconds): 11.551806449999958


### Let's do the same for the 2020 songs

In [59]:
import timeit
start = timeit.default_timer()

# create empty lists where the results are going to be stored
artist_name_20 = []
track_name_20 = []
popularity_20 = []
track_id_20 = []

# the maximum query for Spotify API is 2000 with the max 
for i in range(0,2000,50):
    track_results = sp.search(q='year:2020', type='track', limit=50,offset=i)
    for i, t in enumerate(track_results['tracks']['items']):
        artist_name_20.append(t['artists'][0]['name'])
        track_name_20.append(t['name'])
        track_id_20.append(t['id'])
        popularity_20.append(t['popularity'])
      

stop = timeit.default_timer()
print ('Time to run this code (in seconds):', stop - start)

Time to run this code (in seconds): 6.909060673000113


In [19]:
print('number of elements in the 2019 track_id list:', len(track_id))

number of elements in the 2019 track_id list: 2000


In [60]:
print('number of elements in the 2020 track_id list:', len(track_id_20))

number of elements in the 2020 track_id list: 2000


### Now we need to prepare the data

In [40]:
# 2019
df_tracks_19 = pd.DataFrame({'artist_name':artist_name,'track_name':track_name,'track_id':track_id,'popularity':popularity})
print(df_tracks_19.shape)
df_tracks_19.head()

(2000, 4)


artist_name             track_name                track_id  popularity
0  Roddy Ricch                The Box  0nbXyq5TXYPCO7pr3N8S4I          97
1    SAINt JHN  Roses - Imanbek Remix  7fPuWrlpwDcHm5aHCH5D9t          87
2     Doja Cat                 Say So  3Dv1eDb0MEgF93GpLXlucZ          95
3  Post Malone                Circles  21jGcNKet2qwijlDFuPiPb          93
4     Surfaces            Sunday Best  1Cv1YLb4q0RzL6pybtaMLo          94

In [61]:
# 2020
df_tracks_20 = pd.DataFrame({'artist_name':artist_name_20,'track_name':track_name_20,'track_id':track_id_20,'popularity':popularity_20})
print(df_tracks_20.shape)
df_tracks_20.head()

(2000, 4)


artist_name                    track_name                track_id  \
0      DaBaby  ROCKSTAR (feat. Roddy Ricch)  7ytR5pFWmSjzHJIeQkgog4   
1  THE SCOTTS                    THE SCOTTS  39Yp9wwQiSRIDOvrVg7mbk   
2  The Weeknd               Blinding Lights  0VjIjW4GlUZAMYd2vXMi3b   
3       Drake                  Toosie Slide  127QTOFJsJQp5LbJbu3A1y   
4   Lil Mosey               Blueberry Faygo  6wJYhPfqk3KGhHRG76WzOh   

   popularity  
0          96  
1          97  
2         100  
3          97  
4          93

In [62]:
df_tracks_19.info() 
print('\n')
df_tracks_20.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 4 columns):
artist_name    2000 non-null object
track_name     2000 non-null object
track_id       2000 non-null object
popularity     2000 non-null int64
dtypes: int64(1), object(3)
memory usage: 62.6+ KB


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 4 columns):
artist_name    2000 non-null object
track_name     2000 non-null object
track_id       2000 non-null object
popularity     2000 non-null int64
dtypes: int64(1), object(3)
memory usage: 62.6+ KB


In [56]:
df_tracks_19['popularity'].value_counts()[:10]

67    158
66    131
68    127
65    123
69    121
64    115
63    113
70    110
72    104
71    100
Name: popularity, dtype: int64

### Group the entries by artist_name and track_name and check for duplicates

In [57]:
grouped_19 = df_tracks_19.groupby(['artist_name','track_name'], as_index=True).size()
grouped_19[grouped_19 > 1].count()

56

In [63]:
grouped_20 = df_tracks_20.groupby(['artist_name','track_name'], as_index=True).size()
grouped_20[grouped_20 > 1].count()

240

### There are 56 duplicate entries in 2019 dataset and 240 in 2020 which we will drop

In [64]:
df_tracks_19.drop_duplicates(subset=['artist_name','track_name'], inplace=True)

In [65]:
df_tracks_20.drop_duplicates(subset=['artist_name','track_name'], inplace=True)

In [66]:
# doing the same grouping as before to verify the solution
grouped_after_dropping_19 = df_tracks_19.groupby(['artist_name','track_name'], as_index=True).size()
grouped_after_dropping_19[grouped_after_dropping_19 > 1].count()

0

In [67]:
grouped_after_dropping_20 = df_tracks_20.groupby(['artist_name','track_name'], as_index=True).size()
grouped_after_dropping_20[grouped_after_dropping_20 > 1].count()

0

### This time the results are empty. Another way of checking this:

In [68]:
df_tracks_19[df_tracks_19.duplicated(subset=['artist_name','track_name'],keep=False)].count()

artist_name    0
track_name     0
track_id       0
popularity     0
dtype: int64

In [69]:
df_tracks_19.shape, df_tracks_20.shape

((1943, 4), (1685, 4))

### Get the Audio Features Data

In [70]:
# 2019
# again measuring the time
start = timeit.default_timer()

# empty list, batchsize and the counter for None results
rows = []
batchsize = 100
None_counter = 0

for i in range(0,len(df_tracks_19['track_id']),batchsize):
    batch = df_tracks_19['track_id'][i:i+batchsize]
    feature_results = sp.audio_features(batch)
    for i, t in enumerate(feature_results):
        if t == None:
            None_counter = None_counter + 1
        else:
            rows.append(t)
            
print('Number of 2019 tracks where no audio features were available:',None_counter)

stop = timeit.default_timer()
print ('Time to run this code (in seconds):',stop - start)


Number of 2019 tracks where no audio features were available: 0
Time to run this code (in seconds): 2.7506848000000446


In [71]:
print('number of elements in the 2019 track_id list:', len(rows))

number of elements in the 2019 track_id list: 1943


In [72]:
# 2020
# again measuring the time
start = timeit.default_timer()

# empty list, batchsize and the counter for None results
rows_20 = []
batchsize = 100
None_counter = 0

for i in range(0,len(df_tracks_20['track_id']),batchsize):
    batch = df_tracks_20['track_id'][i:i+batchsize]
    feature_results = sp.audio_features(batch)
    for i, t in enumerate(feature_results):
        if t == None:
            None_counter = None_counter + 1
        else:
            rows_20.append(t)
            
print('Number of 2019 tracks where no audio features were available:',None_counter)

stop = timeit.default_timer()
print ('Time to run this code (in seconds):',stop - start)


Number of 2019 tracks where no audio features were available: 2
Time to run this code (in seconds): 2.2479318939999757


In [73]:
print('number of elements in the 2020 track_id list:', len(rows_20))

number of elements in the 2020 track_id list: 1683


### Load Audio features in a dataframe

In [74]:
# 2019
df_audio_features_19 = pd.DataFrame.from_dict(rows,orient='columns')
print("Shape of the 2019 dataset:", df_audio_features_19.shape)
df_audio_features_19.head()

Shape of the 2019 dataset: (1943, 18)


acousticness                                       analysis_url  \
0        0.1040  https://api.spotify.com/v1/audio-analysis/0nbX...   
1        0.0149  https://api.spotify.com/v1/audio-analysis/7fPu...   
2        0.2560  https://api.spotify.com/v1/audio-analysis/3Dv1...   
3        0.1920  https://api.spotify.com/v1/audio-analysis/21jG...   
4        0.1830  https://api.spotify.com/v1/audio-analysis/1Cv1...   

   danceability  duration_ms  energy                      id  \
0         0.896       196653   0.586  0nbXyq5TXYPCO7pr3N8S4I   
1         0.784       176219   0.721  7fPuWrlpwDcHm5aHCH5D9t   
2         0.787       237893   0.673  3Dv1eDb0MEgF93GpLXlucZ   
3         0.695       215280   0.762  21jGcNKet2qwijlDFuPiPb   
4         0.878       158571   0.525  1Cv1YLb4q0RzL6pybtaMLo   

   instrumentalness  key  liveness  loudness  mode  speechiness    tempo  \
0          0.000000   10    0.7900    -6.687     0       0.0559  116.971   
1          0.004240    8    0.2850    -5.457     1       0.0506  121.966   
2          0.000004   11    0.0904    -4.577     0       0.1580  110.962   
3          0.002440    0    0.0863    -3.497     1       0.0395  120.042   
4          0.000000    5    0.0714    -6.832     1       0.0578  112.022   

   time_signature                                         track_href  \
0               4  https://api.spotify.com/v1/tracks/0nbXyq5TXYPC...   
1               4  https://api.spotify.com/v1/tracks/7fPuWrlpwDcH...   
2               4  https://api.spotify.com/v1/tracks/3Dv1eDb0MEgF...   
3               4  https://api.spotify.com/v1/tracks/21jGcNKet2qw...   
4               4  https://api.spotify.com/v1/tracks/1Cv1YLb4q0Rz...   

             type                                   uri  valence  
0  audio_features  spotify:track:0nbXyq5TXYPCO7pr3N8S4I    0.642  
1  audio_features  spotify:track:7fPuWrlpwDcHm5aHCH5D9t    0.895  
2  audio_features  spotify:track:3Dv1eDb0MEgF93GpLXlucZ    0.786  
3  audio_features  spotify:track:21jGcNKet2qwijlDFuPiPb    0.553  
4  audio_features  spotify:track:1Cv1YLb4q0RzL6pybtaMLo    0.694

In [75]:
# 2020
df_audio_features_20 = pd.DataFrame.from_dict(rows_20,orient='columns')
print("Shape of the 2020 dataset:", df_audio_features_20.shape)
df_audio_features_20.head()

Shape of the 2020 dataset: (1683, 18)


acousticness                                       analysis_url  \
0       0.24700  https://api.spotify.com/v1/audio-analysis/7ytR...   
1       0.23300  https://api.spotify.com/v1/audio-analysis/39Yp...   
2       0.00146  https://api.spotify.com/v1/audio-analysis/0VjI...   
3       0.32100  https://api.spotify.com/v1/audio-analysis/127Q...   
4       0.20700  https://api.spotify.com/v1/audio-analysis/6wJY...   

   danceability  duration_ms  energy                      id  \
0         0.746       181733   0.690  7ytR5pFWmSjzHJIeQkgog4   
1         0.716       165978   0.537  39Yp9wwQiSRIDOvrVg7mbk   
2         0.514       200040   0.730  0VjIjW4GlUZAMYd2vXMi3b   
3         0.834       247059   0.454  127QTOFJsJQp5LbJbu3A1y   
4         0.774       162547   0.554  6wJYhPfqk3KGhHRG76WzOh   

   instrumentalness  key  liveness  loudness  mode  speechiness    tempo  \
0          0.000000   11    0.1010    -7.956     1       0.1640   89.977   
1          0.000000    0    0.1570    -7.648     0       0.0514  129.979   
2          0.000095    1    0.0897    -5.934     1       0.0598  171.005   
3          0.000006    1    0.1140    -9.750     0       0.2010   81.618   
4          0.000000    0    0.1320    -7.909     1       0.0383   99.034   

   time_signature                                         track_href  \
0               4  https://api.spotify.com/v1/tracks/7ytR5pFWmSjz...   
1               4  https://api.spotify.com/v1/tracks/39Yp9wwQiSRI...   
2               4  https://api.spotify.com/v1/tracks/0VjIjW4GlUZA...   
3               4  https://api.spotify.com/v1/tracks/127QTOFJsJQp...   
4               4  https://api.spotify.com/v1/tracks/6wJYhPfqk3KG...   

             type                                   uri  valence  
0  audio_features  spotify:track:7ytR5pFWmSjzHJIeQkgog4    0.497  
1  audio_features  spotify:track:39Yp9wwQiSRIDOvrVg7mbk    0.280  
2  audio_features  spotify:track:0VjIjW4GlUZAMYd2vXMi3b    0.334  
3  audio_features  spotify:track:127QTOFJsJQp5LbJbu3A1y    0.837  
4  audio_features  spotify:track:6wJYhPfqk3KGhHRG76WzOh    0.349

In [76]:
df_audio_features_20.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1683 entries, 0 to 1682
Data columns (total 18 columns):
acousticness        1683 non-null float64
analysis_url        1683 non-null object
danceability        1683 non-null float64
duration_ms         1683 non-null int64
energy              1683 non-null float64
id                  1683 non-null object
instrumentalness    1683 non-null float64
key                 1683 non-null int64
liveness            1683 non-null float64
loudness            1683 non-null float64
mode                1683 non-null int64
speechiness         1683 non-null float64
tempo               1683 non-null float64
time_signature      1683 non-null int64
track_href          1683 non-null object
type                1683 non-null object
uri                 1683 non-null object
valence             1683 non-null float64
dtypes: float64(9), int64(4), object(5)
memory usage: 236.8+ KB


### Drop the features which we will not be use

In [77]:
columns_to_drop = ['analysis_url','track_href','type','uri']
df_audio_features_19.drop(columns_to_drop, axis=1,inplace=True)
df_audio_features_20.drop(columns_to_drop, axis=1,inplace=True)


df_audio_features_19.rename(columns={'id': 'track_id'}, inplace=True)
df_audio_features_20.rename(columns={'id': 'track_id'}, inplace=True)


df_audio_features_19.shape, df_audio_features_20.shape

((1943, 14), (1683, 14))

#### we ended up with df_tracks_19, df_audio_features_19 and df_tracks_20, df_audio_features_20

In [78]:
# merge both dataframes
# the 'inner' method will make sure that we only keep track IDs present in both datasets
df_19 = pd.merge(df_tracks_19,df_audio_features_19,on='track_id',how='inner')
print("Shape of the 2019 dataset:", df_audio_features_19.shape)
df_19.head()

Shape of the 2019 dataset: (1943, 14)


artist_name             track_name                track_id  popularity  \
0  Roddy Ricch                The Box  0nbXyq5TXYPCO7pr3N8S4I          97   
1    SAINt JHN  Roses - Imanbek Remix  7fPuWrlpwDcHm5aHCH5D9t          87   
2     Doja Cat                 Say So  3Dv1eDb0MEgF93GpLXlucZ          95   
3  Post Malone                Circles  21jGcNKet2qwijlDFuPiPb          93   
4     Surfaces            Sunday Best  1Cv1YLb4q0RzL6pybtaMLo          94   

   acousticness  danceability  duration_ms  energy  instrumentalness  key  \
0        0.1040         0.896       196653   0.586          0.000000   10   
1        0.0149         0.784       176219   0.721          0.004240    8   
2        0.2560         0.787       237893   0.673          0.000004   11   
3        0.1920         0.695       215280   0.762          0.002440    0   
4        0.1830         0.878       158571   0.525          0.000000    5   

   liveness  loudness  mode  speechiness    tempo  time_signature  valence  
0    0.7900    -6.687     0       0.0559  116.971               4    0.642  
1    0.2850    -5.457     1       0.0506  121.966               4    0.895  
2    0.0904    -4.577     0       0.1580  110.962               4    0.786  
3    0.0863    -3.497     1       0.0395  120.042               4    0.553  
4    0.0714    -6.832     1       0.0578  112.022               4    0.694

In [79]:
df_20 = pd.merge(df_tracks_20,df_audio_features_20,on='track_id',how='inner')
print("Shape of the 2019 dataset:", df_audio_features_20.shape)
df_20.head()

Shape of the 2019 dataset: (1683, 14)


artist_name                    track_name                track_id  \
0      DaBaby  ROCKSTAR (feat. Roddy Ricch)  7ytR5pFWmSjzHJIeQkgog4   
1  THE SCOTTS                    THE SCOTTS  39Yp9wwQiSRIDOvrVg7mbk   
2  The Weeknd               Blinding Lights  0VjIjW4GlUZAMYd2vXMi3b   
3       Drake                  Toosie Slide  127QTOFJsJQp5LbJbu3A1y   
4   Lil Mosey               Blueberry Faygo  6wJYhPfqk3KGhHRG76WzOh   

   popularity  acousticness  danceability  duration_ms  energy  \
0          96       0.24700         0.746       181733   0.690   
1          97       0.23300         0.716       165978   0.537   
2         100       0.00146         0.514       200040   0.730   
3          97       0.32100         0.834       247059   0.454   
4          93       0.20700         0.774       162547   0.554   

   instrumentalness  key  liveness  loudness  mode  speechiness    tempo  \
0          0.000000   11    0.1010    -7.956     1       0.1640   89.977   
1          0.000000    0    0.1570    -7.648     0       0.0514  129.979   
2          0.000095    1    0.0897    -5.934     1       0.0598  171.005   
3          0.000006    1    0.1140    -9.750     0       0.2010   81.618   
4          0.000000    0    0.1320    -7.909     1       0.0383   99.034   

   time_signature  valence  
0               4    0.497  
1               4    0.280  
2               4    0.334  
3               4    0.837  
4               4    0.349

In [81]:
df_19.info()
print('\n')
df_20.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1943 entries, 0 to 1942
Data columns (total 17 columns):
artist_name         1943 non-null object
track_name          1943 non-null object
track_id            1943 non-null object
popularity          1943 non-null int64
acousticness        1943 non-null float64
danceability        1943 non-null float64
duration_ms         1943 non-null int64
energy              1943 non-null float64
instrumentalness    1943 non-null float64
key                 1943 non-null int64
liveness            1943 non-null float64
loudness            1943 non-null float64
mode                1943 non-null int64
speechiness         1943 non-null float64
tempo               1943 non-null float64
time_signature      1943 non-null int64
valence             1943 non-null float64
dtypes: float64(9), int64(5), object(3)
memory usage: 273.2+ KB


<class 'pandas.core.frame.DataFrame'>
Int64Index: 1683 entries, 0 to 1682
Data columns (total 17 columns):
artist_name        

### check for duplicates

In [82]:
df_19[df_19.duplicated(subset=['artist_name','track_name'],keep=False)]

Empty DataFrame
Columns: [artist_name, track_name, track_id, popularity, acousticness, danceability, duration_ms, energy, instrumentalness, key, liveness, loudness, mode, speechiness, tempo, time_signature, valence]
Index: []

In [83]:
df_20[df_20.duplicated(subset=['artist_name','track_name'],keep=False)]

Empty DataFrame
Columns: [artist_name, track_name, track_id, popularity, acousticness, danceability, duration_ms, energy, instrumentalness, key, liveness, loudness, mode, speechiness, tempo, time_signature, valence]
Index: []

### Now we can create the .csv files for both dataframes

In [84]:
df_19.to_csv('Spotify2019.csv')

In [85]:
df_20.to_csv('Spotify2020.csv')

In [88]:
df.shape, df_19.shape, df_20.shape

((232725, 18), (1943, 17), (1683, 17))

### It's time to merge the dataset together

In [89]:
df.head()

genre        artist_name                        track_name  \
0  Movie     Henri Salvador       C'est beau de faire un Show   
1  Movie  Martin & les fées  Perdu d'avance (par Gad Elmaleh)   
2  Movie    Joseph Williams    Don't Let Me Be Lonely Tonight   
3  Movie     Henri Salvador    Dis-moi Monsieur Gordon Cooper   
4  Movie       Fabien Nataf                         Ouverture   

                 track_id  popularity  acousticness  danceability  \
0  0BRjO6ga9RKCKjfDqeFgWV           0         0.611         0.389   
1  0BjC1NfoEOOusryehmNudP           1         0.246         0.590   
2  0CoSDzoNIKCRs124s9uTVy           3         0.952         0.663   
3  0Gc6TVm52BwZD07Ki6tIvf           0         0.703         0.240   
4  0IuslXpMROHdEPvSl1fTQK           4         0.950         0.331   

   duration_ms  energy  instrumentalness key  liveness  loudness   mode  \
0        99373   0.910             0.000  C#    0.3460    -1.828  Major   
1       137373   0.737             0.000  F#    0.1510    -5.559  Minor   
2       170267   0.131             0.000   C    0.1030   -13.879  Minor   
3       152427   0.326             0.000  C#    0.0985   -12.178  Major   
4        82625   0.225             0.123   F    0.2020   -21.150  Major   

   speechiness    tempo time_signature  valence  
0       0.0525  166.969            4/4    0.814  
1       0.0868  174.003            4/4    0.816  
2       0.0362   99.488            5/4    0.368  
3       0.0395  171.758            4/4    0.227  
4       0.0456  140.576            4/4    0.390

#### Drop 'genre' column from df, we'll not be using it

In [91]:
df = df.drop(columns=['genre'])

In [92]:
df.head()

artist_name                        track_name  \
0     Henri Salvador       C'est beau de faire un Show   
1  Martin & les fées  Perdu d'avance (par Gad Elmaleh)   
2    Joseph Williams    Don't Let Me Be Lonely Tonight   
3     Henri Salvador    Dis-moi Monsieur Gordon Cooper   
4       Fabien Nataf                         Ouverture   

                 track_id  popularity  acousticness  danceability  \
0  0BRjO6ga9RKCKjfDqeFgWV           0         0.611         0.389   
1  0BjC1NfoEOOusryehmNudP           1         0.246         0.590   
2  0CoSDzoNIKCRs124s9uTVy           3         0.952         0.663   
3  0Gc6TVm52BwZD07Ki6tIvf           0         0.703         0.240   
4  0IuslXpMROHdEPvSl1fTQK           4         0.950         0.331   

   duration_ms  energy  instrumentalness key  liveness  loudness   mode  \
0        99373   0.910             0.000  C#    0.3460    -1.828  Major   
1       137373   0.737             0.000  F#    0.1510    -5.559  Minor   
2       170267   0.131             0.000   C    0.1030   -13.879  Minor   
3       152427   0.326             0.000  C#    0.0985   -12.178  Major   
4        82625   0.225             0.123   F    0.2020   -21.150  Major   

   speechiness    tempo time_signature  valence  
0       0.0525  166.969            4/4    0.814  
1       0.0868  174.003            4/4    0.816  
2       0.0362   99.488            5/4    0.368  
3       0.0395  171.758            4/4    0.227  
4       0.0456  140.576            4/4    0.390

### Concat all three datasets to get a single .csv file

In [120]:
df_final = pd.concat([df, df_19, df_20], ignore_index=True)

In [121]:
print(df_final.shape)
df_final.head()

(236351, 17)


artist_name                        track_name  \
0     Henri Salvador       C'est beau de faire un Show   
1  Martin & les fées  Perdu d'avance (par Gad Elmaleh)   
2    Joseph Williams    Don't Let Me Be Lonely Tonight   
3     Henri Salvador    Dis-moi Monsieur Gordon Cooper   
4       Fabien Nataf                         Ouverture   

                 track_id  popularity  acousticness  danceability  \
0  0BRjO6ga9RKCKjfDqeFgWV           0         0.611         0.389   
1  0BjC1NfoEOOusryehmNudP           1         0.246         0.590   
2  0CoSDzoNIKCRs124s9uTVy           3         0.952         0.663   
3  0Gc6TVm52BwZD07Ki6tIvf           0         0.703         0.240   
4  0IuslXpMROHdEPvSl1fTQK           4         0.950         0.331   

   duration_ms  energy  instrumentalness key  liveness  loudness   mode  \
0        99373   0.910             0.000  C#    0.3460    -1.828  Major   
1       137373   0.737             0.000  F#    0.1510    -5.559  Minor   
2       170267   0.131             0.000   C    0.1030   -13.879  Minor   
3       152427   0.326             0.000  C#    0.0985   -12.178  Major   
4        82625   0.225             0.123   F    0.2020   -21.150  Major   

   speechiness    tempo time_signature  valence  
0       0.0525  166.969            4/4    0.814  
1       0.0868  174.003            4/4    0.816  
2       0.0362   99.488            5/4    0.368  
3       0.0395  171.758            4/4    0.227  
4       0.0456  140.576            4/4    0.390

### Double check for duplicates

In [122]:
grouped = df_final.groupby(['artist_name','track_name'], as_index=True).size()
grouped[grouped > 1].count()

35448

In [123]:
df_final.shape

(236351, 17)

In [130]:
df_final[df_final.duplicated(subset=['artist_name','track_name'],keep=False)].count()

artist_name         92127
track_name          92127
track_id            92127
popularity          92127
acousticness        92127
danceability        92127
duration_ms         92127
energy              92127
instrumentalness    92127
key                 92127
liveness            92127
loudness            92127
mode                92127
speechiness         92127
tempo               92127
time_signature      92127
valence             92127
dtype: int64

In [131]:
df_final[df_final.duplicated()].count()

artist_name         41669
track_name          41669
track_id            41669
popularity          41669
acousticness        41669
danceability        41669
duration_ms         41669
energy              41669
instrumentalness    41669
key                 41669
liveness            41669
loudness            41669
mode                41669
speechiness         41669
tempo               41669
time_signature      41669
valence             41669
dtype: int64

In [132]:
df_final = df_final.drop_duplicates()

In [134]:
df_final.shape

(194682, 17)

### Create the .csv file

In [136]:
df_final.to_csv('df_final.csv')